aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -57,6 +57,58 @@ TORCH_META_FUNC(mm)(const Tensor & self, const Tensor & mat2) {
   "The input tensor must be a matrix with size ", self.sizes()[0], "x", mat2.sizes()[1], ", but got a ", result.dim(),
   "-D tensor with size ", result.sizes()[0], "x", result.sizes()[1]);
 }
+
+template <typename Meta>
+void common_checks_baddbmm_bmm(Meta& meta, const Tensor& batch1, const Tensor& batch2, const Scalar& beta, const Scalar& alpha, bool is_bmm, const c10::optional<Tensor>& self_baddbmm = nullopt) {
+  TORCH_CHECK(batch1.dim() == 3, "batch1 must be a 3D tensor");
+  TORCH_CHECK(batch2.dim() == 3, "batch2 must be a 3D tensor");
+
+  const auto batch1_sizes = batch1.sizes();
+  const auto batch2_sizes = batch2.sizes();
+
+  int64_t bs = batch1_sizes[0];
+  int64_t contraction_size = batch1_sizes[2];
+  int64_t res_rows = batch1_sizes[1];
+  int64_t res_cols = batch2_sizes[2];
+
+  TORCH_CHECK(batch2_sizes[0] == bs && batch2_sizes[1] == contraction_size);
+
+  auto& result = meta.maybe_get_output(0);
+  // 'set_output' does not resize for in-place calls
+  meta.set_output({bs, res_rows, res_cols}, batch1.options());
+  const auto result_sizes = result.sizes();
+  // Error is raised if called from in-place overload with incorrect shape
+  TORCH_CHECK(result_sizes[0] == bs && result_sizes[1] == res_rows && result_sizes[2] == res_cols);
```

Comment:
Nit: an error message here would be nice!

Reply:
You don't have to spell it out individually, ArrayRef has equality defined. This also solves a bug where you can OOB access if the dimensionality differs. Also another one for https://github.com/pytorch/pytorch/issues/57827

Reply:
@ysiraichi maybe you can fix the underlying issue, so we don't have to keep pasting this in? :)

Reply:
Right, to me figuring out the error message gets tricky here (but I've still made an effort to be as correct as possible, in the recent commit) because an error is raised if:

1. `out` tensor is passed with an incorrect shape.
2. (in-place call), input tensor is passed with an incorrect shape.

I've added an error message in the recent commit, please let me know if it sounds good or if you have any suggestions. However, we'll not reach this error frequently, a lot of times it will fail at `expand_size`:

```python
In [8]: M = torch.randn(1, 3, 2)

In [9]: M.baddbmm_(batch1, batch2)
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-9-a9e665d29b16> in <module>
----> 1 M.baddbmm_(batch1, batch2)

RuntimeError: The expanded size of the tensor (5) must match the existing size (2) at non-singleton dimension 2.  Target sizes: [10, 3, 5].  Tensor sizes: [1, 3, 2]
```

The only time when we'll get this error is for in-place calls, as expected:

```python
In [12]: M = torch.randn(10, 1, 5)

# batch1 size: (10, 3, 4) and batch2 size: (10, 4, 5)
In [13]: M.baddbmm_(batch1, batch2).size()
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-13-ffc510337fae> in <module>
----> 1 M.baddbmm_(batch1, batch2).size()

RuntimeError: Expected an output tensor with shape [10, 3, 5] but got shape [10, 1, 5]
```

Hope this error message looks good now. For completeness, I've also added a similar error message for `self_sizes` check in the same function. :)

Reply:
> You don't have to spell it out individually, ArrayRef has equality defined. This also solves a bug where you can OOB access if the dimensionality differs. Also another one for #57827

Sorry, I just noticed this.  This should be fixed now, defined `output_size` (`IntArrayRef`) since it's used twice.

Also added error messages here:

```python
# batch2 is of incorrect size
In [7]: M.baddbmm_(batch1, batch2)
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-7-a9e665d29b16> in <module>
----> 1 M.baddbmm_(batch1, batch2)

RuntimeError: Expected size for first two dimensions of batch2 tensor to be: [10, 4] but got: [10, 5].
```

[Relevant commit](https://github.com/pytorch/pytorch/pull/64805/commits/86a408220e20f5b8d45accbbbd5b55a74ac30ae2)